File: backend/src/schemas/inspection.py
from pydantic import BaseModel
from datetime import datetime
from typing import List, Dict, Optional, Any


class InspectionTaskCreate(BaseModel):
    """创建巡检任务"""
    name: str
    scan_type: Optional[str] = "full"
    is_enabled: Optional[bool] = True
    cron_expr: Optional[str] = "0 */4 * * *"
    
    # 扫描目标（三选一）
    target_type: Optional[str] = "all_devices"
    site_id: Optional[int] = None
    ip_range: Optional[str] = None
    
    # SNMP配置
    snmp_community: Optional[str] = "public"
    snmp_version: Optional[str] = "v2c"
    snmp_timeout: Optional[int] = 3
    snmp_retries: Optional[int] = 1
    
    # TCP探测配置
    tcp_ports: Optional[List[int]] = [22, 80, 443, 445, 3389]
    tcp_timeout_ms: Optional[int] = 2000
    
    # 并发控制
    max_concurrent: Optional[int] = 50
    
    # 告警配置
    alert_on_offline: Optional[bool] = True
    alert_on_new_device: Optional[bool] = True
    alert_on_fingerprint_change: Optional[bool] = True


class InspectionTaskUpdate(BaseModel):
    """更新巡检任务"""
    name: Optional[str] = None
    scan_type: Optional[str] = None
    is_enabled: Optional[bool] = None
    cron_expr: Optional[str] = None
    
    # 扫描目标
    target_type: Optional[str] = None
    site_id: Optional[int] = None
    ip_range: Optional[str] = None
    
    # SNMP配置
    snmp_community: Optional[str] = None
    snmp_version: Optional[str] = None
    snmp_timeout: Optional[int] = None
    snmp_retries: Optional[int] = None
    
    # TCP探测配置
    tcp_ports: Optional[List[int]] = None
    tcp_timeout_ms: Optional[int] = None
    
    # 并发控制
    max_concurrent: Optional[int] = None
    
    # 告警配置
    alert_on_offline: Optional[bool] = None
    alert_on_new_device: Optional[bool] = None
    alert_on_fingerprint_change: Optional[bool] = None


class InspectionTaskResponse(BaseModel):
    """巡检任务响应"""
    id: int
    name: str
    scan_type: str
    is_enabled: bool
    cron_expr: str
    
    # 扫描目标
    target_type: str
    site_id: Optional[int]
    ip_range: Optional[str]
    
    # SNMP配置
    snmp_community: str
    snmp_version: str
    snmp_timeout: int
    snmp_retries: int
    
    # TCP探测配置
    tcp_ports: List[int]
    tcp_timeout_ms: int
    
    # 并发控制
    max_concurrent: int
    
    # 告警配置
    alert_on_offline: bool
    alert_on_new_device: bool
    alert_on_fingerprint_change: bool
    
    # 执行记录
    last_run_at: Optional[datetime]
    last_run_status: Optional[str]
    
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


class InspectionResultResponse(BaseModel):
    """巡检结果响应"""
    id: int
    task_id: int
    scan_type: str
    trigger: str
    operator: str
    status: str
    total_targets: int
    online_count: int
    offline_count: int
    new_device_count: int
    change_count: int
    error_message: Optional[str]
    started_at: datetime
    finished_at: Optional[datetime]
    duration_seconds: Optional[float]
    
    class Config:
        from_attributes = True


class DeviceFingerprintResponse(BaseModel):
    """设备指纹响应"""
    id: int
    ip_address: str
    device_id: Optional[int]
    sys_descr: Optional[str]
    sys_name: Optional[str]
    sys_object_id: Optional[str]
    sys_location: Optional[str]
    vendor: Optional[str]
    last_seen_online: Optional[datetime] = None
    last_full_scan_at: Optional[datetime] = None
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


class InspectionDeviceResultResponse(BaseModel):
    """单设备巡检结果响应"""
    id: int
    result_id: int
    ip_address: str
    device_id: Optional[int]
    is_online: bool
    detection_method: str
    open_ports: Optional[List[int]]
    
    # SNMP采集结果
    sys_descr: Optional[str]
    sys_name: Optional[str]
    sys_object_id: Optional[str]
    sys_up_time: Optional[int]
    sys_location: Optional[str]
    vendor: Optional[str]
    cpu_usage: Optional[float]
    memory_usage: Optional[float]
    
    # 变更检测
    is_new_device: bool
    has_fingerprint_change: bool
    change_detail: Optional[Dict[str, List[str]]]
    
    # 执行信息
    scan_duration_ms: int
    error_message: Optional[str]
    scanned_at: datetime
    
    class Config:
        from_attributes = True


class AlertCountResponse(BaseModel):
    """告警统计响应"""
    total: int = 0
    unresolved: int = 0
    new_device: int = 0
    missing_device: int = 0
    changed_device: int = 0

File: frontend/src/api/inspection.ts
import api from './axios'

export interface InspectionTask {
  id: number
  name: string
  scan_type: string
  is_enabled: boolean
  cron_expr: string
  target_type: string
  site_id: number | null
  ip_range: string | null
  snmp_community: string
  snmp_version: string
  snmp_timeout: number
  snmp_retries: number
  tcp_ports: number[]
  tcp_timeout_ms: number
  max_concurrent: number
  alert_on_offline: boolean
  alert_on_new_device: boolean
  alert_on_fingerprint_change: boolean
  last_run_at: string | null
  last_run_status: string | null
  created_at: string
  updated_at: string | null
}

export interface InspectionTaskCreate {
  name: string
  scan_type?: string
  is_enabled?: boolean
  cron_expr?: string
  target_type?: string
  site_id?: number
  ip_range?: string
  snmp_community?: string
  snmp_version?: string
  snmp_timeout?: number
  snmp_retries?: number
  tcp_ports?: number[]
  tcp_timeout_ms?: number
  max_concurrent?: number
  alert_on_offline?: boolean
  alert_on_new_device?: boolean
  alert_on_fingerprint_change?: boolean
}

export interface InspectionTaskUpdate {
  name?: string
  scan_type?: string
  is_enabled?: boolean
  cron_expr?: string
  target_type?: string
  site_id?: number
  ip_range?: string
  snmp_community?: string
  snmp_version?: string
  snmp_timeout?: number
  snmp_retries?: number
  tcp_ports?: number[]
  tcp_timeout_ms?: number
  max_concurrent?: number
  alert_on_offline?: boolean
  alert_on_new_device?: boolean
  alert_on_fingerprint_change?: boolean
}

export interface InspectionRecord {
  id: number
  task_id: number
  scan_type: string
  trigger: string
  operator: string
  status: string
  total_targets: number
  online_count: number
  offline_count: number
  new_device_count: number
  change_count: number
  error_message: string | null
  started_at: string
  finished_at: string | null
  duration_seconds: number | null
}

export interface DeviceFingerprint {
  id: number
  ip_address: string
  device_id: number | null
  sys_descr: string | null
  sys_name: string | null
  sys_object_id: string | null
  sys_location: string | null
  vendor: string | null
  last_seen_online: string
  last_full_scan_at: string
  updated_at: string | null
}

export interface InspectionDeviceResult {
  id: number
  result_id: number
  ip_address: string
  device_id: number | null
  is_online: boolean
  detection_method: string
  open_ports: number[] | null
  sys_descr: string | null
  sys_name: string | null
  sys_object_id: string | null
  sys_up_time: number | null
  sys_location: string | null
  vendor: string | null
  cpu_usage: number | null
  memory_usage: number | null
  is_new_device: boolean
  has_fingerprint_change: boolean
  change_detail: Record<string, string[]> | null
  scan_duration_ms: number
  error_message: string | null
  scanned_at: string
}

export interface AlertCount {
  total: number
  unresolved: number
  new_device: number
  missing_device: number
  changed_device: number
}

export interface ExecuteResponse {
  result_id: number
}

// 巡检任务相关API
export async function getTasks(): Promise<InspectionTask[]> {
  const response = await api.get('/inspection/tasks')
  return response.data
}

export async function getTask(taskId: number): Promise<InspectionTask> {
  const response = await api.get(`/inspection/tasks/${taskId}`)
  return response.data
}

export async function createTask(data: InspectionTaskCreate): Promise<InspectionTask> {
  const response = await api.post('/inspection/tasks', data)
  return response.data
}

export async function updateTask(taskId: number, data: InspectionTaskUpdate): Promise<InspectionTask> {
  const response = await api.put(`/inspection/tasks/${taskId}`, data)
  return response.data
}

export async function deleteTask(taskId: number): Promise<void> {
  await api.delete(`/inspection/tasks/${taskId}`)
}

export async function toggleTask(taskId: number): Promise<{ message: string; is_enabled: boolean }> {
  const response = await api.patch(`/inspection/tasks/${taskId}/toggle`)
  return response.data
}

export async function executeTask(taskId: number, scanType?: string): Promise<ExecuteResponse> {
  const params: Record<string, unknown> = {}
  if (scanType) params.scan_type = scanType
  const response = await api.post(`/inspection/tasks/${taskId}/run`, undefined, { params })
  return response.data
}

// 巡检记录相关API
export async function getRecords(taskId?: number, scanType?: string, status?: string, limit = 20): Promise<InspectionRecord[]> {
  const params: Record<string, unknown> = { limit }
  if (taskId) params.task_id = taskId
  if (scanType) params.scan_type = scanType
  if (status) params.status = status
  const response = await api.get('/inspection/results', { params })
  return response.data
}

export async function getRecord(recordId: number): Promise<InspectionRecord> {
  const response = await api.get(`/inspection/results/${recordId}`)
  return response.data
}

export async function getDeviceResults(recordId: number, filters?: {
  is_online?: boolean
  is_new_device?: boolean
  has_fingerprint_change?: boolean
}): Promise<InspectionDeviceResult[]> {
  const params: Record<string, unknown> = {}
  if (filters?.is_online !== undefined) params.is_online = filters.is_online
  if (filters?.is_new_device !== undefined) params.is_new_device = filters.is_new_device
  if (filters?.has_fingerprint_change !== undefined) params.has_fingerprint_change = filters.has_fingerprint_change
  const response = await api.get(`/inspection/results/${recordId}/devices`, { params })
  return response.data
}

// 设备指纹相关API
export async function getFingerprints(vendor?: string, ip?: string): Promise<DeviceFingerprint[]> {
  const params: Record<string, unknown> = {}
  if (vendor) params.vendor = vendor
  if (ip) params.ip = ip
  const response = await api.get('/inspection/fingerprints', { params })
  return response.data
}

export async function getFingerprint(ip: string): Promise<DeviceFingerprint> {
  const response = await api.get(`/inspection/fingerprints/${ip}`)
  return response.data
}

// 告警统计
export async function getAlertCount(): Promise<AlertCount> {
  const response = await api.get('/inspection/alerts/count')
  return response.data
}
